### ISSUE TITLE
RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:294 : Failed to meet rendezvous 'loss_reduce': Socket closed (14)

### ISSUE THREAD

**Issue by jacekpoplawski** [pytorch/xla#2387]
hello, this is one of times I see following error on Kaggle:

```
Exception in device=TPU:6: tensorflow/compiler/xla/xla_client/mesh_service.cc:294 : Failed to meet rendezvous 'loss_reduce': Socket closed (14)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 231, in _start_fn
    fn(gindex, *args)
  File "<ipython-input-4-7212b5ecb6e6>", line 6, in _mp_fn
    training(model, config)
  File "../input/common-20200730-2043/common/siim_isic_training.py", line 74, in training
    result = epoch(m_o_s, dataloaders, device, criterion)
Exception in device=TPU:5: tensorflow/compiler/xla/xla_client/mesh_service.cc:294 : Failed to meet rendezvous 'loss_reduce': Socket closed (14)
  File "../input/common-20200730-2043/common/siim_isic_epoch.py", line 35, in epoch
    train_result = train_step(config_step)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 231, in _start_fn
    fn(gindex, *args)
  File "<ipython-input-4-7212b5ecb6e6>", line 6, in _mp_fn
    training(model, config)
  File "../input/common-20200730-2043/common/siim_isic_train_step.py", line 72, in train_step
    reduced_loss = xm.mesh_reduce('loss_reduce', loss, reduce_fn)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 727, in mesh_reduce
    xdata = rendezvous(tag, bio.getvalue())
  File "../input/common-20200730-2043/common/siim_isic_training.py", line 74, in training
    result = epoch(m_o_s, dataloaders, device, criterion)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 679, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:294 : Failed to meet rendezvous 'loss_reduce': Socket closed (14)
  File "../input/common-20200730-2043/common/siim_isic_epoch.py", line 35, in epoch
    train_result = train_step(config_step)
  File "../input/common-20200730-2043/common/siim_isic_train_step.py", line 72, in train_step
    reduced_loss = xm.mesh_reduce('loss_reduce', loss, reduce_fn)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 727, in mesh_reduce
    xdata = rendezvous(tag, bio.getvalue())
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 679, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:294 : Failed to meet rendezvous 'loss_reduce': Socket closed (14)

---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-4-7212b5ecb6e6> in <module>
      9 
     10 xmp.spawn(_mp_fn, args=(config,), nprocs=8,
---> 11           start_method='fork')

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    294         join=join,
    295         daemon=daemon,
--> 296         start_method=start_method)
    297 
    298 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    156 
    157     # Loop on join until it returns True or raises an exception.
--> 158     while not context.join():
    159         pass
    160 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    106                 raise Exception(
    107                     "process %d terminated with signal %s" %
--> 108                     (error_index, name)
    109                 )
    110             else:

Exception: process 0 terminated with signal SIGSEGV

```

my current training step loop is:
```

    for ii, (x, meta, extra, target) in enumerate(tk0):    
            
        if (device):
            x, meta, target, extra = x.to(device), meta.to(device), target.to(device), extra.to(device)
        
        if (train_mode):
            optimizer.zero_grad()
            
        output = model(x, meta, extra)
        
        loss = criterion(output, target.unsqueeze(1).type_as(output))

        if (train_mode):                
            if (use_tpu):
                loss.backward()                
                xm.optimizer_step(optimizer)
            else:
                loss.backward()                
                optimizer.step()                          
                
            scheduler.step()                         
        
        y_preds.append(output.cpu().detach().numpy())
        
        if (use_tpu):
            reduced_loss = xm.mesh_reduce('loss_reduce', loss, reduce_fn)       
        else:
            reduced_loss = loss

        losses.append([reduced_loss.item(), output.shape[0]])
  
        y_true.append(target.unsqueeze(1).type_as(output).cpu().detach().numpy())

```

I use following version:

```

!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py  --version 20200530 

```

**Comment by JackCaoG** [comment]
I think the real error is the `Exception: process 0 terminated with signal SIGSEGV`, does your code consistently fail? Could you provide us a small repo so we can better debug this segfault issue?

**Comment by jacekpoplawski** [comment]
I can't - this is Kaggle competition. However, do you mean the error is not related to reduce but this is just random segfault in my code? So one of cores crashed and others can't continue and stop on reduce, right?

**Comment by JackCaoG** [comment]
`rendezvous` is used to sync between all processes, if one of the process exit early(in your case segfault), other process will wait for a while and then report Failed to meet rendezvous.

**Comment by jacekpoplawski** [comment]
Is there any way I can get info about this segfault? As I wrote above it is random, even when my code is fully reproducible.

**Comment by JackCaoG** [comment]
Debugging segfault remotely is a bit challenging, but if you can dump the c++ stack when it segfault again it will be super helpful.  Currently we don't know this comes from pytorch, pytorch/xla or somewhere else. I usually run the code with gdb to get that but I am not sure if that is doable on Kaggle (`gdb --args python xxxx`).  There should be an easiler way to dump the C++ stack on Kaggle but I didn't find it by doing a quick search.

**Comment by jacekpoplawski** [comment]
I am running same code multiple times with different efficientnet architecture, it crashed only once today. Single run is about hour. So it's very rare and random. Can it be related to the version of xla I downloaded (see above)?

**Comment by JackCaoG** [comment]
You are using a relatively old version of the pt/xla (5/30's nightly) but I am not very familiar with our Kaggle wheel. @jysohn23 Would it be enough to just update the  `--version` to be a more recent day? Is the server version default to be the most recent version or do we also pinned it to an older tfrt?

**Comment by jysohn23** [comment]
Yeah you can just change the `version` flag in the setup cell. That script also takes care of selecting the matching TPU TFRC runtime.

**Comment by jacekpoplawski** [comment]
I used this version because it was recommended in another issue here, which version should I use? I was thinking latest one can be unstable

**Comment by JackCaoG** [comment]
I don't think we have any fix merge in this week, so any date of this week should work. We usually recommend user to use the most recent version of pt/xla as it is likely to be more stable and with more features. Which issue are you referring(about the version 05/30)?

**Comment by jacekpoplawski** [comment]
This one https://github.com/pytorch/xla/issues/2364
so should I just put "nightly" as a version?

**Comment by jacekpoplawski** [comment]
BTW you said about remotely debugging - it is possible to use TPU not remotely? I mean can anyone buy one for home or use something else instead to execute this code?

**Comment by JackCaoG** [comment]
Oh, when I said remote I meant I don't have the access to the vm/notebook that actually segfault. I don't think there is a way to purchase TPU right now. Yes "nightly" is good.

**Comment by jacekpoplawski** [comment]
thanks I will switch to nightly tomorrow, today I want to finish testing with same code to avoid any differences in results
